cdc/sinkv2/eventsink/mq/mq_sink.go

```diff
@@ -14,24 +14,122 @@
 package mq
 
 import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/contextutil"
 	"github.com/pingcap/tiflow/cdc/model"
+	mqv1 "github.com/pingcap/tiflow/cdc/sink/mq"
+	"github.com/pingcap/tiflow/cdc/sink/mq/codec"
+	"github.com/pingcap/tiflow/cdc/sink/mq/dispatcher"
+	"github.com/pingcap/tiflow/cdc/sink/mq/manager"
 	"github.com/pingcap/tiflow/cdc/sinkv2/eventsink"
+	"github.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq/producer"
+	"github.com/pingcap/tiflow/pkg/config"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/util"
+	"go.uber.org/zap"
 )
 
+// newProducerImpl specifies the build method for the producer.
+var newProducerImpl producer.NewProducerFunc = producer.NewMockProducer
+
 // Assert EventSink[E event.TableEvent] implementation
-var _ eventsink.EventSink[*model.RowChangedEvent] = (*Sink)(nil)
+var _ eventsink.EventSink[*model.RowChangedEvent] = (*sink)(nil)
+
+// sink is the mq sink.
+// It will send the events to the MQ system.
+type sink struct {
+	// id indicates this sink belongs to which processor(changefeed).
+	id model.ChangeFeedID
+	// role indicates this sink used for what.
+	role util.Role
+	// protocol indicates the protocol used by this sink.
+	protocol config.Protocol
+
+	worker *worker
+	// eventRouter used to route events to the right topic and partition.
+	eventRouter *dispatcher.EventRouter
+	// topicManager used to manage topics.
+	// It is also responsible for creating topics.
+	topicManager manager.TopicManager
+
+	// encoderBuilder builds encoder for the sink.
+	encoderBuilder codec.EncoderBuilder
+}
+
+func newSink(ctx context.Context,
+	producer producer.Producer,
+	topicManager manager.TopicManager,
+	eventRouter *dispatcher.EventRouter,
+	encoderConfig *codec.Config,
+	errCh chan error,
+) (*sink, error) {
+	changefeedID := contextutil.ChangefeedIDFromCtx(ctx)
```

Comment:
Could you pass changefeed id in function param explicitly?

Reply:
We seem to use it this way in a lot of places. The place that calls it will presumably only be able to get it that way.